src/secp256k1.c

```diff
@@ -472,6 +522,9 @@ int secp256k1_ecdsa_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature
                 break;
             }
             secp256k1_scalar_set_b32(&non, nonce32, &overflow);
+            if (nonce_tweak32 != NULL) {
+                secp256k1_scalar_add(&non, &non, &nonce_tweak);
```

Comment:
This is not a sign-to-contract commitment, but it works for the anti nonce covertchan because the host already has the nonce. I think it would be much better if `ecdsa_sign` would generally support sign to contract commitments and not only this protocol. Sign to contract would be `R = R1 + H(R1, k2)*G`.

Reply:
Thanks a lot for the great review! 

I can try to make a general sign to contract ecdsa sign function.

Peeking at your Schnorr implementation of it, I am unsure about a few things:

1. There is `secp256k1_s2c_opening`, which seems to return the original `R1` from the sign function. Why is this needed when there is already a function to produce the same commitment (`secp256k1_(schnorrsig|ecdsa)_anti_nonce_sidechan_client_commit`) beforehand?

2. What is `nonce_is_negated` and when is it set? [Here](https://github.com/bitcoin-core/secp256k1/pull/590/files#diff-313ca26f0048bc16a608709915d0111eR206) the check is `if (!secp256k1_fe_is_quad_var(&r.y)) {`, but I don't know what this means. Does it apply for ecdsa?

In the ecdsa case, do I need to add all of that, or is just adding `H(R1, k2)` instead of `k2` sufficient?

Reply:
I know pushed a commit which simply adds `H(R1, k2)` to the nonce to produce `R`. Seemed less intrusive and more efficient than computing `R2` and adding it to `R1` to get `R`.

Reply:
Question 2. is resolved by reading [BIP-Schnorr](https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki). Unless I misunderstand, it seems that it does not apply to ecdsa.

Reply:
To question 1: True, but if you want sign-to-contract outside of the anti-nonce-sidechan context then it's much simpler to just have a single opening structure that contains the negated nonce and original nonce and that you can just give to s2c_verify.